src/coreclr/jit/lclmasks.cpp

```diff
@@ -0,0 +1,480 @@
+            JITDUMP("Local %s V%02d at [%06u] has %s conversion. ", isLocalStore ? "store" : "var", lclOp->GetLclNum(),
+                    m_compiler->dspTreeID(lclOp), hasConversion ? "mask" : "no");
+            weight.UpdateWeight(isLocalStore, hasConversion, bbWeight);
+
+            // Cache the simd type data of the conversion.
+            if (hasConversion)
+            {
+                assert(convertOp != nullptr);
+                weight.CacheSimdTypes(convertOp);
+            }
+
+            // Update the table.
+            weightsTable->Set(lclOp->GetLclNum(), weight, LclMasksWeightTable::Overwrite);
+
+            foundConversions |= hasConversion;
+        }
+
+        return fgWalkResult::WALK_CONTINUE;
+    }
+
+    bool foundConversions = false;
+
+private:
+    weight_t             bbWeight;
+    LclMasksWeightTable* weightsTable;
+};
+
+//-----------------------------------------------------------------------------
+// LclMasksUpdateVisitor: tree visitor to remove conversion to masks for uses of LCL
+//
+class LclMasksUpdateVisitor final : public GenTreeVisitor<LclMasksUpdateVisitor>
+{
+public:
+    enum
+    {
+        DoPostOrder       = true,
+        UseExecutionOrder = true
+    };
+
+    LclMasksUpdateVisitor(Compiler* compiler, Statement* stmt, LclMasksWeightTable* weightsTable)
+        : GenTreeVisitor<LclMasksUpdateVisitor>(compiler)
+        , stmt(stmt)
+        , weightsTable(weightsTable)
+    {
+    }
+
+    Compiler::fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTreeLclVarCommon* lclOp            = nullptr;
+        bool                 isLocalStore     = false;
+        bool                 isLocalUse       = false;
+        bool                 addConversion    = false;
+        bool                 removeConversion = false;
+
+        if ((*use)->OperIs(GT_STORE_LCL_VAR) && (*use)->AsLclVarCommon()->Data()->OperIsConvertMaskToVector())
+        {
+            // Found
+            //      use:STORE_LCL_VAR(ConvertMaskToVector(x))
+            lclOp            = (*use)->AsLclVarCommon();
+            isLocalStore     = true;
+            removeConversion = true;
+        }
+        else if ((*use)->OperIs(GT_STORE_LCL_VAR) && !(*use)->AsLclVarCommon()->Data()->OperIsConvertMaskToVector())
+        {
+            // Found
+            //      use:STORE_LCL_VAR(x)
+            lclOp         = (*use)->AsLclVarCommon();
+            isLocalStore  = true;
+            addConversion = true;
+        }
+        else if ((*use)->OperIsConvertVectorToMask() && (*use)->AsHWIntrinsic()->Op(2)->OperIs(GT_LCL_VAR))
+        {
+            // Found
+            //      user(use:ConvertVectorToMask(LCL_VAR(x)))
+            lclOp            = (*use)->AsHWIntrinsic()->Op(2)->AsLclVarCommon();
+            isLocalUse       = true;
+            removeConversion = true;
+        }
+        else if ((*use)->OperIs(GT_LCL_VAR) && (!user->OperIsConvertVectorToMask()))
```

Comment:
```suggestion
        else if ((*use)->OperIs(GT_LCL_VAR) && (user != nullptr) && !user->OperIsConvertVectorToMask())
```
Top-level `GT_LCL_VAR` can exist in some odd cases.

Reply:
switched to:
```
else if ((*use)->OperIs(GT_LCL_VAR) && ((user == nullptr) || !user->OperIsConvertVectorToMask()))
```

As top level `GT_LCL_VAR` is valid here.